url47087446e9_kellrott_SparqlScape_tgz-pJ8-org_json_TestJ8/src/org/sparqlscape/SparqlInterface.java

```diff
@@ -0,0 +1,392 @@
+		nodeTypeScrollPane.setPreferredSize(new Dimension(300, 200) );
+		c.gridx=1;
+		panel.add( nodeTypeScrollPane, c );
+
+		Button okButton = new Button("OK");
+		okButton.addActionListener( new ActionListener() {					
+			@Override
+			public void actionPerformed(ActionEvent e) {
+				frame.setVisible(false);
+			}
+		}); 
+		c.gridx=0;
+		c.gridy = 3;
+		panel.add( okButton, c );
+
+		frame.add( panel );
+		frame.setLocationRelativeTo(null);
+		frame.setSize(600,300);
+		frame.setModal(true);		
+		frame.pack();
+		frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);		
+		frame.setVisible(true);
+
+		Map<String, Boolean> nodeTypeHash = new HashMap<String, Boolean>();
+		for (Object o : nodeTypeList.getSelectedValues() ) {
+			String s = (String)o;
+			nodeTypeHash.put(s, true);
+		}
+		List<LinkTriple> outList = new LinkedList<LinkTriple>();
+
+		for ( LinkTriple link : inList ) {
+			if ( nodeTypeHash.containsKey( link.pred.getType() ) 
+					&& nodeTypeHash.containsKey( link.subj.getType() ) 
+			) {
+				outList.add(link);
+			}
+		}
+
+		return outList.toArray( new LinkTriple[0] );
+	}
+
+	public Boolean dialogDone;
+	public LinkTriple [] NodeSearch( Component parent, final List<Map<String,Object>> nodeAttributes, final String idCol ) {
+
+		JPanel panel = new JPanel();
+
+		panel.setLayout( new GridBagLayout() );
+		GridBagConstraints c = new GridBagConstraints();
+
+		c.gridy = 1;
+		panel.add( new Label("SPARQL Search"), c );
+		c.gridx = 1;
+		final TextArea sparqlQuery = new TextArea(6, 50);
+		panel.add( sparqlQuery, c );
+
+		final Map<LinkTriple, Integer> nameList = new HashMap<LinkTriple, Integer>();
+		final JDialog frame = new JDialog();
+
+		c.gridx = 0;
+		c.gridy = 3;
+		panel.add( new Label("LinkCount:"), c);
+		c.gridx = 1;
+		final Label linkCountLabel = new Label("0");
+		panel.add( linkCountLabel, c);		
+
+		c.gridx = 0;
+		c.gridy = 6;
+		panel.add( new JLabel("<html><p>SPARQL</p><p>INPUT node id: {id}</p><p>OUTPUT  ?src  ?edge ?dst</p></html>"), c  );
+
+		Button searchButton = new Button("Search Selected"); 
+		searchButton.addActionListener( new ActionListener() {
+			@Override
+			public void actionPerformed(ActionEvent e) {
+				String searchBase = sparqlQuery.getText() ;
+				nameList.clear();
+				try {
+					Client sClient = new Client(new URL(endpointURL));
+					for ( Map<String,Object> attrs : nodeAttributes ) {
+						for ( String key : attrs.keySet() ) {
+							Pattern idRegex = Pattern.compile( "\\{" + key + "\\}" );
+							String searchStr = idRegex.matcher(searchBase).replaceAll( attrs.get(key).toString() );		
+							Results results = sClient.select(searchStr);
+							String colName = results.getResultVars().get(0);
+							for ( Row row : results ) {
+								System.out.println( row.getResource(colName).getURI() );
+								nameList.put( new LinkTriple( new SparqlData(attrs.get(key).toString(), true), new SparqlData("sparqlLink", true), row.getResource(colName) ), 1 );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.